elfi/testbench/testbench.py

```diff
@@ -0,0 +1,301 @@
+    def _resolve_test_type(self):
+        self._set_default_test_type()
+        self._resolve_observations()
+        self._resolve_reference_parameters()
+
+    def _set_default_test_type(self):
+        self.description = {
+            'observations_available': self.observations is not None,
+            'reference_parameters_available': self.reference_parameter is not None,
+            'reference_posterior_available': self.reference_posterior is not None
+            }
+
+    def _resolve_reference_parameters(self):
+        if self.description['reference_parameters_available']:
+            for keys, values in self.reference_parameter.items():
+                self.reference_parameter[keys] = np.repeat(
+                    values,
+                    repeats=self.repetitions
+                    )
+
+        elif not self.description['observations_available']:
+            seed = self._get_seeds(n_rep=1)
+            self.reference_parameter = self.model.generate(
+                batch_size=self.repetitions,
+                outputs=self.model.parameter_names,
+                seed=seed[0])
+
+    def _resolve_observations(self):
+        if self.description['observations_available']:
+            self.observations = np.repeat(
+                self.observations,
+                repeats=self.repetitions,
+                axis=0)
+        else:
+            seed = self._get_seeds(n_rep=1)
+            self.observations = self.model.generate(
+                with_values=self.reference_parameter,
+                outputs=self.simulator_name,
+                batch_size=self.repetitions,
+                seed=seed[0])[self.simulator_name]
+
+    def add_method(self, new_method):
+        """Add a new method to the testbench.
+
+        Parameters
+        ----------
+        new_method : TestbenchMethod
+            An inference method as a TestbenchMethod.
+
+        """
+        logger.info('Adding {} to testbench.'.format(new_method.attributes['name']))
+        self.method_list.append(new_method)
+        self.method_seed_list.append(self._get_seeds(n_rep=self.repetitions))
+
+    def run(self):
+        """Run Testbench."""
+        self.testbench_results = []
+        for method_index, method in enumerate(self.method_list):
+            logger.info('Running {} in testbench.'.format(method.attributes['name']))
+
+            if self.progress_bar:
+                self.progress_bar.reinit_progressbar(reinit_msg=method.attributes['name'])
+
+            self.testbench_results.append(
+                self._repeat_inference(method, self.method_seed_list[method_index])
+                )
+
+    def _repeat_inference(self, method, seed_list):
+        repeated_result = []
+        model = self.model.copy()
+        for i in np.arange(self.repetitions):
+            if self.progress_bar:
+                self.progress_bar.update_progressbar(i + 1, self.repetitions)
+
+            model.observed[self.simulator_name] = self.observations[i]
```

Comment:
could use `model.observed[self.simulator_name] = np.atleast_2d(self.observations[i])` so that observations are in the same format as simulator output and we can be sure that summaries calculation works as expected

Reply:
Sounds like a very good idea and added! (Even though there shouldn't be a compatibility issue here?)

Reply:
i had one model where i had not taken into account that the input to summaries calculation could be 1d :) it is my mistake but i assume it could happen to someone else as well.